apollo-client/src/main/java/com/ctrip/framework/apollo/internals/K8sConfigMapConfigRepository.java

```diff
@@ -0,0 +1,301 @@
+        if (StringUtils.isBlank(configMapName)) {
+            throw new IllegalArgumentException("ConfigMap name cannot be null");
+        }
+        if (kubernetesManager.checkConfigMapExist(k8sNamespace, configMapName)) {
+            return;
+        }
+        // Create an empty configmap, write the new value in the update event
+        Transaction transaction = Tracer.newTransaction("Apollo.ConfigService", "createK8sConfigMap");
+        transaction.addData("configMapName", configMapName);
+        try {
+            kubernetesManager.createConfigMap(k8sNamespace, configMapName, null);
+            transaction.setStatus(Transaction.SUCCESS);
+        } catch (Throwable ex) {
+            Tracer.logEvent("ApolloConfigException", ExceptionUtil.getDetailMessage(ex));
+            transaction.setStatus(ex);
+            throw new ApolloConfigException("Create configmap failed!", ex);
+        } finally {
+            transaction.complete();
+        }
+    }
+
+    @Override
+    public Properties getConfig() {
+        if (configMapProperties == null) {
+            sync();
+        }
+        Properties result = propertiesFactory.getPropertiesInstance();
+        result.putAll(configMapProperties);
+        logger.info("configmap properties: {}", configMapProperties);
+        return result;
+    }
+
+    /**
+     * Update the memory when the configuration center changes
+     *
+     * @param upstreamConfigRepository the upstream repo
+     */
+    @Override
+    public void setUpstreamRepository(ConfigRepository upstreamConfigRepository) {
+        // 设置上游数据源
+        if (upstreamConfigRepository == null) {
+            return;
+        }
+        //clear previous listener
+        if (upstream != null) {
+            upstream.removeChangeListener(this);
+        }
+        upstream = upstreamConfigRepository;
+        upstreamConfigRepository.addChangeListener(this);
+    }
+
+    @Override
+    public ConfigSourceType getSourceType() {
+        return sourceType;
+    }
+
+    /**
+     * Sync the configmap
+     */
+    @Override
+    protected void sync() {
+        // Chain recovery, first read from upstream data source
+        boolean syncFromUpstreamResultSuccess = trySyncFromUpstream();
+
+        if (syncFromUpstreamResultSuccess) {
+            return;
+        }
+
+        Transaction transaction = Tracer.newTransaction("Apollo.ConfigService", "syncK8sConfigMap");
+        Throwable exception = null;
+        try {
+            configMapProperties = loadFromK8sConfigMap();
+            sourceType = ConfigSourceType.CONFIGMAP;
+            transaction.setStatus(Transaction.SUCCESS);
+        } catch (Throwable ex) {
+            Tracer.logEvent("ApolloConfigException", ExceptionUtil.getDetailMessage(ex));
+            transaction.setStatus(ex);
+            exception = ex;
+            throw new ApolloConfigException("Load config from Kubernetes ConfigMap failed!", ex);
```

Comment:
This logic seems flawed. If this line executes, then line 193 and lines 199 to 203 are unnecessary.